File: pages/mypage/mypage_account_page.py
# pages/mypage/mypage_account_page.py
# 마이페이지 > 계정 관리 — 기본 정보 전용 Page 클래스
# FHC-073(이름 변경) / FHC-074(비밀번호 변경) / FHC-075(프로모션 알림) / FHC-076(선호 언어)

import time

from config.selenium_imports import By, EC

from pages.mypage.mypage_page import MyPage


class MyPage05(MyPage):

    # ========== Locators — FHC-073: 이름 변경 ==========
    NAME_INPUT = (By.CSS_SELECTOR, "input[name='fullname']")
    SAVE_SUCCESS_TOAST = (By.XPATH,
        "//*[contains(text(),'저장되었습니다') or contains(text(),'Saved successfully')"
        " or contains(text(),'saved')]"
    )
    # 이름/비밀번호 인라인 편집 완료 버튼
    COMPLETE_BUTTON = (By.XPATH,
        "//button[@type='submit' and contains(@class,'MuiLoadingButton-root')]"
    )

    # ========== Locators — FHC-074: 비밀번호 변경 ==========
    CURRENT_PW_INPUT = (By.CSS_SELECTOR, "input[autocomplete='current-password']")
    NEW_PW_INPUT     = (By.CSS_SELECTOR, "input[name='newPassword']")
    CONFIRM_PW_INPUT = (By.CSS_SELECTOR, "input[name='confirmPassword']")

    # ========== Locators — FHC-075: 프로모션 알림 ==========
    PROMOTION_TOGGLE = (By.XPATH,
        "//input[@type='checkbox' and ("
        "@name='marketing' or @name='promotion' or @name='promotionAlarm'"
        ")]"
        " | //*[contains(text(),'프로모션 알림 받기')]/ancestor::label//input[@type='checkbox']"
        " | //*[contains(text(),'프로모션 알림 받기')]/following-sibling::*//input[@type='checkbox']"
    )

    # ========== FHC-073: 이름 변경 ==========

    def click_name_edit(self):
        """이름 옆 편집 아이콘 클릭 (JS — SVG 네임스페이스 문제 우회)"""
        result = self.driver.execute_script("""
            var s = document.querySelectorAll('svg[data-testid="EditOutlinedIcon"]');
            if (s.length > 0 && s[0].closest('button')) {
                s[0].closest('button').click(); return 'ok';
            }
            var b = document.querySelectorAll('button.MuiIconButton-root');
            if (b.length > 0) { b[0].click(); return 'fallback'; }
            return 'not-found';
        """)
        if result == 'not-found':
            raise Exception("이름 편집 버튼을 찾을 수 없습니다")
        self.logger.info(f"이름 편집 버튼 클릭 완료 ({result})")

    def enter_name(self, name: str):
        name_input = self.wait.until(EC.visibility_of_element_located(self.NAME_INPUT))
        name_input.clear()
        name_input.send_keys(name)
        self.logger.info(f"이름 입력 완료: {name}")

    def save_name(self):
        self.js_click(
            self.wait.until(EC.element_to_be_clickable(self.COMPLETE_BUTTON))
        )
        self.logger.info("이름 저장(완료) 버튼 클릭 완료")

    def is_save_success_toast_displayed(self) -> bool:
        try:
            self.wait.until(EC.presence_of_element_located(self.SAVE_SUCCESS_TOAST))
            return True
        except Exception:
            return False

    # ========== FHC-074: 비밀번호 변경 ==========

    def click_password_edit(self):
        """비밀번호 옆 편집 아이콘 클릭 (JS — SVG 네임스페이스 문제 우회)"""
        result = self.driver.execute_script("""
            var s = document.querySelectorAll('svg[data-testid="EditOutlinedIcon"]');
            if (s.length >= 4 && s[3].closest('button')) {
                s[3].closest('button').click(); return 'ok';
            }
            var b = document.querySelectorAll('button.MuiIconButton-root');
            if (b.length >= 4) { b[3].click(); return 'fallback'; }
            return 'not-found';
        """)
        if result == 'not-found':
            raise Exception("비밀번호 편집 버튼을 찾을 수 없습니다")
        self.logger.info(f"비밀번호 편집 버튼 클릭 완료 ({result})")

    def change_password(self, current_pw: str, new_pw: str):
        """비밀번호 변경 (current_pw → new_pw, 동일 비밀번호 불가)"""
        curr = self.wait.until(EC.visibility_of_element_located(self.CURRENT_PW_INPUT))
        curr.clear()
        curr.send_keys(current_pw)

        new = self.wait.until(EC.visibility_of_element_located(self.NEW_PW_INPUT))
        new.clear()
        new.send_keys(new_pw)

        confirm = self.wait.until(EC.visibility_of_element_located(self.CONFIRM_PW_INPUT))
        confirm.clear()
        confirm.send_keys(new_pw)

        self.js_click(
            self.wait.until(EC.element_to_be_clickable(self.COMPLETE_BUTTON))
        )
        self.logger.info(f"비밀번호 변경 완료: {current_pw} → {new_pw}")
        time.sleep(1)

    # ========== FHC-075: 프로모션 알림 토글 ==========

    def get_promotion_state(self) -> bool:
        checkbox = self.wait.until(
            EC.presence_of_element_located(self.PROMOTION_TOGGLE)
        )
        state = checkbox.is_selected()
        self.logger.info(f"프로모션 알림 현재 상태: {'ON' if state else 'OFF'}")
        return state

    def toggle_promotion(self):
        checkbox = self.wait.until(
            EC.presence_of_element_located(self.PROMOTION_TOGGLE)
        )
        self.driver.execute_script("arguments[0].click();", checkbox)
        time.sleep(0.5)
        self.logger.info("프로모션 알림 토글 클릭 완료")

File: pages/MyPageAccountPage.ts
import { Page, expect } from '@playwright/test'
import { BasePage } from './BasePage'

const ACCOUNT_URL  = 'https://accounts.elice.io/accounts/members/account'
const LANGUAGE_URL = 'https://accounts.elice.io/accounts/members/language'
const LOGIN_URL =
  'https://accounts.elice.io/accounts/signin/me' +
  '?continue_to=https%3A%2F%2Fqaproject.elice.io%2Fai-helpy-chat' +
  '&lang=ko-KR&org=qaproject'

export class MyPageAccountPage extends BasePage {
  constructor(page: Page) { super(page) }

  get nameInput()          { return this.page.locator("input[name='fullname']") }
  get currentPwInput()     { return this.page.locator("input[autocomplete='current-password']") }
  get newPwInput()         { return this.page.locator("input[name='newPassword']") }
  get confirmPwInput()     { return this.page.locator("input[name='confirmPassword']") }
  get completeButton()     { return this.page.locator("button[type='submit'].MuiLoadingButton-root") }
  get saveSuccessToast()   { return this.page.getByText(/저장되었습니다|Saved successfully|saved/) }
  get promotionToggle()    {
    // Selenium과 동일한 다중 선택자 (marketing / promotion / promotionAlarm / 텍스트 기반)
    return this.page.locator([
      "input[type='checkbox'][name='marketing']",
      "input[type='checkbox'][name='promotion']",
      "input[type='checkbox'][name='promotionAlarm']",
      "label:has-text('프로모션 알림 받기') input[type='checkbox']",
    ].join(', ')).first()
  }
  get languageSelect()     { return this.page.locator("div#mui-component-select-locale, div[aria-haspopup='listbox']") }

  async navigateToAccount(): Promise<void> {
    await this.page.goto(ACCOUNT_URL)
    // signin으로 리디렉션 시 accounts 세션이 없는 것 — 상위에서 skip 처리
    await expect(this.page).toHaveURL(/members\/account/, { timeout: 15000 })
    // 콘텐츠 완전 로드 대기 (로딩 중 편집 버튼 미노출 방지)
    await this.page.waitForLoadState('networkidle', { timeout: 10000 }).catch(() => null)
  }

  async login(email: string, password: string): Promise<void> {
    await this.page.goto(LOGIN_URL)
    await this.page.locator('[name="loginId"]').fill(email)
    await this.page.locator('[name="password"]').fill(password)
    await this.page.getByRole('button', { name: '로그인' }).click()
    await expect(this.page).toHaveURL(/ai-helpy-chat/, { timeout: 30000 })
  }

  async clickNameEdit(): Promise<void> {
    // Selenium의 click_name_edit(): SVG 기반 우선, fallback으로 MuiIconButton
    await this.page.evaluate(() => {
      const svgs = document.querySelectorAll('svg[data-testid="EditOutlinedIcon"]')
      if (svgs.length > 0) {
        const btn = (svgs[0] as Element).closest('button') as HTMLElement | null
        if (btn) { btn.click(); return }
      }
      const btns = document.querySelectorAll<HTMLButtonElement>('button.MuiIconButton-root')
      if (btns[0]) btns[0].click()
    })
    // 편집 폼(input)이 나타날 때까지 대기
    await this.nameInput.waitFor({ timeout: 10000 })
  }

  async enterName(name: string): Promise<void> {
    await this.nameInput.clear()
    await this.nameInput.fill(name)
  }

  async saveName(): Promise<void> {
    await this.completeButton.click()
  }

  async clickPasswordEdit(): Promise<void> {
    await this.page.evaluate(() => {
      const btns = document.querySelectorAll<HTMLButtonElement>('button.MuiIconButton-root')
      if (btns[3]) btns[3].click()
    })
  }

  async changePassword(currentPw: string, newPw: string): Promise<void> {
    await this.currentPwInput.fill(currentPw)
    await this.newPwInput.fill(newPw)
    await this.confirmPwInput.fill(newPw)
    await this.completeButton.click()
    await this.page.waitForTimeout(1000)
  }

  async getPromotionState(): Promise<boolean> {
    return await this.promotionToggle.isChecked()
  }

  async togglePromotion(): Promise<void> {
    await this.page.evaluate(() => {
      const cb = document.querySelector<HTMLInputElement>(
        "input[type='checkbox'][name='marketing'], input[type='checkbox'][name='promotion']"
      )
      if (cb) cb.click()
    })
  }

  async changeLanguage(lang: string): Promise<void> {
    // Selenium: navigate_to_language() → change_language() — submit 버튼 없음, 드롭다운 선택 시 자동 저장
    await this.page.goto(LANGUAGE_URL)
    await expect(this.page).toHaveURL(/members\/language/, { timeout: 10000 })
    await this.page.waitForLoadState('networkidle', { timeout: 10000 }).catch(() => null)
    // 현재 값과 관계없이 저장이 트리거되도록 중간 언어로 먼저 변경
    const mid = lang === 'en-US' ? 'ko-KR' : 'en-US'
    await this.languageSelect.first().click()
    await this.page.locator(`li[data-value='${mid}']`).first().click()
    await this.page.waitForTimeout(400)
    await this.languageSelect.first().click()
    await this.page.locator(`li[data-value='${lang}']`).first().click()
    await this.page.waitForTimeout(500)
  }

  async isSaveSuccessToastDisplayed(): Promise<void> {
    await expect(this.saveSuccessToast.first()).toBeVisible({ timeout: 10000 })
  }

  async isSavedSuccessfullyDisplayed(): Promise<void> {
    await expect(this.page.getByText(/Saved successfully|저장되었습니다/).first()).toBeVisible({ timeout: 10000 })
  }
}
